src/MiddlewareListener.php

```diff
@@ -146,6 +143,16 @@ private function createPipeFromSpec(
                 throw InvalidMiddlewareException::fromMiddlewareName($middlewareName);
             }
 
+            // Decorate double-pass middleware
+            if (is_callable($middlewareToBePiped)
+                && (
+                    ! $middlewaresToBePiped instanceof MiddlewareInterface
+                    || ! method_exists($middlewaresToBePiped, 'process')
```

Comment:
These changes should covered by unit tests.

Reply:
A new test `testSuccessfullyDispatchesDoublePassMiddleware`?

Reply:
For example.

Reply:
I've added the test, but as I come to think about the if condition, doesn't it have to be `x && !y && !z` instead of `x && (!y || !z)`.. the first one makes more sense than the second one. Otherwise middlewares that are not an instance of `MiddlewareInterface`, but have the `process` method would be decorated, as well. Hm...